src/Microsoft.DncEng.SecretManager/SecretTypes/ADApplication.cs

```diff
@@ -0,0 +1,127 @@
+
+            DateTime rotateOn = expiresOn.AddDays(-15);
+
+            if (parameters.KustoConnectionStrings != null)
+            {
+                var kustoConnectionStringsSecrets = parameters.KustoConnectionStrings.Split(',', StringSplitOptions.RemoveEmptyEntries);
+                foreach (var kustoConnectionStringSecret in kustoConnectionStringsSecrets)
+                {
+                    await RollKustoConnectionString(context, kustoConnectionStringSecret.Trim(), appId, appSecret, expiresOn, rotateOn);
+                }
+            }
+
+            return new List<SecretData> {
+                    new SecretData(appId, DateTimeOffset.MaxValue, DateTimeOffset.MaxValue),
+                    new SecretData(appSecret, expiresOn, rotateOn) };
+        }
+
+        private async Task RollKustoConnectionString(RotationContext context, string secretName, string appId, string appSecret, DateTime expiresOn, DateTime rotateOn)
+        {
+            string connectionString = await context.GetSecretValue(new SecretReference(secretName));
+
+            if (string.IsNullOrEmpty(connectionString))
+            {
+                _console.WriteLine($"Configure a new Kusto connection string {secretName}");
+                string dataSource = await _console.PromptAndValidateAsync("Data source",
+                                                "Expecting URI",
+                                                l => Uri.TryCreate(l, UriKind.Absolute, out _));
+                string initialCatalog = await _console.PromptAndValidateAsync("Initial catalog",
+                                                "Expecting non empty string",
+                                                l => !string.IsNullOrWhiteSpace(l));
+                string additionalParameters = await _console.PromptAsync("Enter additional parameters: ");
+
+                if (!string.IsNullOrWhiteSpace(additionalParameters))
+                    additionalParameters = ";" + additionalParameters;
+
+                connectionString = $"Data Source={dataSource};Initial Catalog={initialCatalog};AAD Federated Security=True;Application Client Id={appId};{ApplicationKey}{appSecret}{additionalParameters}";
+            }
+            else
+            {
+                string[] parts = connectionString.Split(';', StringSplitOptions.RemoveEmptyEntries);
+                for (int i = 0; i < parts.Length; i++)
+                {
+                    if (parts[i].StartsWith(ApplicationKey, true, CultureInfo.InvariantCulture))
+                        parts[i] = ApplicationKey + appSecret;
+                }
+                connectionString = string.Join(';', parts);
+            }
+
+            _console.WriteLine($"Storing the Kusto connection string {secretName} with value {connectionString}");
+
+            await context.SetSecretValue(secretName, new SecretValue(connectionString, context.GetValues(), rotateOn, expiresOn));
```

Comment:
This is a bit problematic. This won't really work correctly. The 'verify' command has no way to know that this secret also produces another secret name, so any references to this kusto connection string in the settings.json will be considered errors.

I don't even thing we really need to talk about "kusto connection strings" at all. The consuming service can just construct it from the AD id and secret. We can put "kusto connection string" explicitly in the manifest, but the better way would be another secret type that just references this one, similar to how the sas-uri types all reference the storage key secret.

Reply:
Oh right, it would be a problem. I added a new secret kusto connection string which is linked to application secret.
Then I found that our app loads secrets in random order and we need to have it ordered. I needed to add topological sort into sync command to get it working.

Reply:
I don't think we need to go as far as making a complicated topological sort. Rotating an AD application secret shouldn't invalidate the old value until it has been rotated a second time. Thats how the storage keys work. Any single key will be valid for 2 rotation periods, so the derived secrets will expire well before the storage key does. As long as the rotation interval for the derived secret is not longer than the interval for the source secret, things will stay consistent.